### PR TITLE
Add ARP functions.

### PR DESCRIPTION
**Pull request by robshakir**
```
commit 462b221da0e4b3a7e436f43b506a671fc55a43e5
Author: Rob Shakir <robjs@google.com>
Date:   Wed Aug 17 18:14:26 2022 -0700

    Add integration test against real Linux netlink.

     * (A) integration/..,
       - Add an integration test against the netlink library that can run on a
         real Linux machine to check netlink accessor implementation.

commit debbb43a75b52861baaf307ca493e457e635b151
Author: Rob Shakir <robjs@google.com>
Date:   Wed Aug 17 16:52:11 2022 -0700

    Add ARPSubscribe mechanism.
    
     * (M) intf/accessor_linux.go
      - Add a mechanism to subscribe to ARP entries that are populated
        in the kernel.
     * (M) intf/intf(_test)?.go
      - Add mechanism that subscribes to the ARP table independently
        of the underlying implementation.
```

### CODE REVIEW

File: intf/intf.go [hunk 1]
```diff
@@ -35,12 +36,17 @@ type Interface struct {
...
+	return fmt.Sprintf("%s (index %d, MAC: %s)", i.Name, i.Index, i.MAC)
+}
+
 // ARPEntry is a representation of an ARP entry on the underlying system.
 type ARPEntry struct {
 	// IP is the IP address of the neighbour.
 	IP net.IP
 	// Interface is the name of the interface on which the ARP entry was learnt.
-	Interface string
+	Interface *Interface
```
Comment (akrentsel) [inline intf/intf.go:49]: github doesn't let me comment on other parts on the code...this comment is meant for line 70. The comment there is unfinished?

File: intf/intf.go [hunk 2]
```diff
@@ -79,9 +85,12 @@ type NetworkAccessor interface {
 	// AddressAdd adds address ip to the interface name.
 	AddInterfaceIP(name string, ip *net.IPNet) error
 	// ARPList lists the set of ARP neighbours on the system.
-	ARPList() ([]*ARPNeigh, error)
+	ARPList() ([]*ARPEntry, error)
 	// ARPSubscribe writes changes to the ARP table to the channel updates, and uses
-	// done to indicate that the subscription is complete.
+	// done to indicate that the subscription is complete. The ARPSubscribe function
```
Comment (akrentsel) [inline intf/intf.go:90]: for my understanding - I assume not-blocking because it is a "subscribe" function, and that naming implies non-blocking?

then if we had wanted to make a blocking version, would the naming be "ARPGet"?

File: intf/intf.go [hunk 3]
```diff
@@ -154,3 +168,56 @@ func AddIP(intf string, addr *net.IPNet) error {
...
+	result := make(chan ARPUpdate, 1)
+	done := make(chan struct{})
+
+	go func(updates chan ARPUpdate, done chan struct{}) {
+		for {
+			select {
+			case upd := <-updates:
+				if upd.Type == ARPAdd {
+					if addr.Equal(upd.Neigh.IP) {
+						result <- upd
```
Comment (akrentsel) [inline intf/intf.go:197]: how do closures work for golang? is it fine that you're writing to a channel that you didn't pass a pointer in for? Why did you need to pass updates and done into the function, but not resukt?

File: intf/intf.go [hunk 4]
```diff
@@ -154,3 +168,56 @@ func AddIP(intf string, addr *net.IPNet) error {
...
+
+	if err := accessor.ARPSubscribe(updates, done); err != nil {
+		return nil, fmt.Errorf("cannot subscribe to ARP updates, err: %v", err)
+	}
+
+	select {
+	case n := <-result:
+		done <- struct{}{}
+		return n.Neigh.MAC, nil
+	case <-ctx.Done():
```
Comment (akrentsel) [inline intf/intf.go:214]: is this case trigger from doing ctx.WithTimeout()? I assume after 30 seconds, an empty struct is put into the ctx.Done() channel?

File: intf/intf.go [hunk 5]
```diff
@@ -154,3 +168,56 @@ func AddIP(intf string, addr *net.IPNet) error {
...
+	go func(updates chan ARPUpdate, done chan struct{}) {
+		for {
+			select {
+			case upd := <-updates:
+				if upd.Type == ARPAdd {
+					if addr.Equal(upd.Neigh.IP) {
+						result <- upd
+					}
+				}
+			case <-done:
```
Comment (akrentsel) [inline intf/intf.go:200]: so the usage pattern in the ARPSubscribe will be that it'll ARP out for the given IP, and upon receiving responses, will write them to the updates channel? then at some point write to done to tell AwaitARP that no more responses are coming? 

Why do we not return if `addr.Equal(upd.Neigh.IP)`, meaning we successfully found the MAC address of the requested IP?

**Review by akrentsel — APPROVED** [review]

File: intf/intf.go (hunk 1 above)
Comment (robshakir) [reply intf/intf.go:49]: Thanks - finished this sentence (just needed a period).

File: intf/intf.go (hunk 4 above)
Comment (robshakir) [reply intf/intf.go:214]: That's correct -- `ctx.Done()` returns a channel that is written to when that channel becomes complete - e.g., the timeout fires, or the cancellation function is called.

File: intf/intf.go (hunk 3 above)
Comment (robshakir) [reply intf/intf.go:197]: It is fine to be writing to a channel this way - we can do both: refer to the channel outside of our scope, or hand it in to make it within the scope. In this case these two things are equivalent. 

Updated to make sure that this is consistent in this code.

File: intf/intf.go (hunk 2 above)
Comment (robshakir) [reply intf/intf.go:90]: It could be blocking and mean that the caller has to put it into a goroutine if they want to not block, but it seemed more reasonable to just make it a non-blocking function itself (since this is the pattern that I expect most callers to use).

File: intf/intf.go (hunk 5 above)
Comment (robshakir) [reply intf/intf.go:200]: It doesn't trigger ARP, but it will wait until ARP for that address is resolved. It writes the updates that it receives for that IP to the channel so that the caller can do something with that now-resolved ARP entry.

It doesn't write to `done`, but rather uses that to determine whether it should keep listening - so we get this goroutine to return when we write to `done` on line 212.

I had originally had this as a more generic listener, where it would allow you to continue to listen, you're right that this can be simplified to just return the goroutine when it gets the first update, given the caller does that.

We still write to `done` on 212 because that tells the accessor's underlying `ARPSubscribe` to exit.

**Comment by robshakir** [comment]
Alex - will merge here, but we can discuss the comments next week if that works for you :-)
